File: src/OrderBookTest.java
import org.junit.Test;

import java.util.ArrayList;

import static org.junit.Assert.*;

public class OrderBookTest {

    @Test
    public void placeBuy() {
        ArrayList<Transaction> ts = new ArrayList<>();

        OrderBook b = new OrderBook("GOOG");
        b.placeBuy(1, 100, "David", 1);
        b.placeBuy(5, 100, "Brian", 2);
        b.placeBuy(20, 100, "Arjun", 3);
        b.placeBuy(4, 101, "Kevin", 4);

        b.placeSell(2, 101, "Andrew", 5);
        ts.add(new Transaction(
                "Andrew", "Kevin", "GOOG",
                2, 101, -1));

        b.placeSell(4, 100, "Bob", 6);
        ts.add(new Transaction(
                "Bob", "Kevin", "GOOG",
                2, 100, -1
        ));
        ts.add(new Transaction(
                "Bob", "David", "GOOG",
                1, 100, -1
        ));
        ts.add(new Transaction(
                "Bob", "Brian", "GOOG",
                1, 100, -1
        ));
        assertEquals(100, b.getMarketPrice(), .1);

        b.placeSell(5, 98, "Jake", 7);
        ts.add(new Transaction(
                "Jake", "Brian", "GOOG",
                4, 98, -1
        ));
        ts.add(new Transaction(
                "Jake", "Arjun", "GOOG",
                1, 98, -1
        ));
        assertEquals(98, b.getMarketPrice(), .1);

        ArrayList<Transaction> list = b.getTransactions();
        for (int i = 0; i < ts.size(); i++) {
            assertTrue(transEqual(list.get(i), ts.get(i)));
        }
    }

    @Test
    public void placeSell() {
        ArrayList<Transaction> ts = new ArrayList<>();

        OrderBook b = new OrderBook("GOOG");
        b.placeSell(1, 100, "David", 1);
        b.placeSell(5, 100, "Brian", 2);
        b.placeSell(20, 100, "Arjun", 3);
        b.placeSell(4, 99, "Kevin", 4);

        b.placeBuy(2, 99, "Andrew", 5);
        ts.add(new Transaction(
                "Kevin", "Andrew", "GOOG",
                2, 99, -1));

        b.placeBuy(4, 100, "Bob", 6);
        ts.add(new Transaction(
                "Kevin", "Bob", "GOOG",
                2, 100, -1
        ));
        ts.add(new Transaction(
                "David", "Bob", "GOOG",
                1, 100, -1
        ));
        ts.add(new Transaction(
                "Brian", "Bob", "GOOG",
                1, 100, -1
        ));
        assertEquals(100, b.getMarketPrice(), .1);

        b.placeBuy(5, 103, "Jake", 7);
        ts.add(new Transaction(
                "Brian", "Jake", "GOOG",
                4, 103, -1
        ));
        ts.add(new Transaction(
                "Arjun", "Jake", "GOOG",
                1, 103, -1
        ));
        assertEquals(103, b.getMarketPrice(), .1);

        ArrayList<Transaction> list = b.getTransactions();
        for (int i = 0; i < ts.size(); i++) {
            assertTrue(transEqual(list.get(i), ts.get(i)));
        }
    }

    @Test
    public void getMarketPrice() {}

    public boolean transEqual(Transaction a, Transaction b) {
        return a.security.equals(b.security) && a.amt == b.amt
                && a.price == b.price && a.from.equals(b.from)
                && a.to.equals(b.to);
    }
}

File: orderbook_test.go
package main

import (
	"testing"
)

func TestPlaceBuy(t *testing.T) {
	var ts []*Transaction

	b := NewOrderBook("GOOG")
	b.PlaceBuy(1, 100, "David", 1)
	b.PlaceBuy(5, 100, "Brian", 2)
	b.PlaceBuy(20, 100, "Arjun", 3)
	b.PlaceBuy(4, 101, "Kevin", 4)
	if b.buys.Len() != 4 {
		t.Fatal("Length incorrect")
	}

	b.PlaceSell(2, 101, "Andrew", 5)
	ts = append(ts, &Transaction{"Andrew", "Kevin", "GOOG", 2, 101, -1})

	b.PlaceSell(4, 100, "Bob", 6)
	ts = append(ts,
		&Transaction{"Bob", "Kevin", "GOOG", 2, 100, -1},
		&Transaction{"Bob", "David", "GOOG", 1, 100, -1},
		&Transaction{"Bob", "Brian", "GOOG", 1, 100, -1},
	)
	if b.MarketPrice() != 100 {
		t.Fatal("Market price failed. Expected: 100 Got:", b.MarketPrice())
	}

	b.PlaceSell(5, 98, "Jake", 7)
	ts = append(ts,
		&Transaction{"Jake", "Brian", "GOOG", 4, 98, -1},
		&Transaction{"Jake", "Arjun", "GOOG", 1, 98, -1},
	)
	if b.MarketPrice() != 98 {
		t.Fatal("Market price failed. Expected: 98 Got:", b.MarketPrice())
	}

	for i := range ts {
		if !transEqual(b.transactions[i], ts[i]) {
			t.Fatal("Failed on ", i, "Expected:", ts[i], "Got:", b.transactions[i])
		}
	}
}

func TestPlaceSell(t *testing.T) {
	var ts []*Transaction

	b := NewOrderBook("GOOG")
	b.PlaceSell(1, 100, "David", 1)
	b.PlaceSell(5, 100, "Brian", 2)
	b.PlaceSell(20, 100, "Arjun", 3)
	b.PlaceSell(4, 99, "Kevin", 4)
	if b.sells.Len() != 4 {
		t.Fatal("Length incorrect")
	}

	b.PlaceBuy(2, 99, "Andrew", 5)
	ts = append(ts, &Transaction{"Kevin", "Andrew", "GOOG", 2, 99, -1})

	b.PlaceBuy(4, 100, "Bob", 6)
	ts = append(ts,
		&Transaction{"Kevin", "Bob", "GOOG", 2, 100, -1},
		&Transaction{"David", "Bob", "GOOG", 1, 100, -1},
		&Transaction{"Brian", "Bob", "GOOG", 1, 100, -1},
	)
	if b.MarketPrice() != 100 {
		t.Fatal("Market price failed. Expected: 100 Got:", b.MarketPrice())
	}

	b.PlaceBuy(5, 103, "Jake", 7)
	ts = append(ts,
		&Transaction{"Brian", "Jake", "GOOG", 4, 103, -1},
		&Transaction{"Arjun", "Jake", "GOOG", 1, 103, -1},
	)
	if b.MarketPrice() != 103 {
		t.Fatal("Market price failed. Expected: 98 Got:", b.MarketPrice())
	}

	for i := range ts {
		if !transEqual(b.transactions[i], ts[i]) {
			t.Fatal("Failed on ", i, "Expected:", ts[i], "Got:", b.transactions[i])
		}
	}
}

func transEqual(a *Transaction, b *Transaction) bool {
	return a.Security == b.Security && a.Amt == b.Amt && a.Price == b.Price && a.From == b.From && a.To == a.To
}
